crates/core/tedge_api/src/entity_store.rs

```diff
@@ -184,14 +179,63 @@ impl EntityStore {
             twin_data: Map::new(),
         };
 
-        Some(EntityStore {
+        let message_log = MqttMessageLog::new(log_path)?;
+
+        let mut entity_store = EntityStore {
+            mqtt_schema,
             main_device: main_device.topic_id.clone(),
             entities: HashMap::from([(main_device.topic_id.clone(), metadata)]),
             entity_id_index: HashMap::from([(entity_id, main_device.topic_id)]),
             external_id_mapper: Box::new(external_id_mapper_fn),
             external_id_validator_fn: Box::new(external_id_validator_fn),
             default_service_type,
-        })
+            message_log,
+        };
+
+        entity_store.load_from_message_log()?;
+
+        Ok(entity_store)
+    }
+
+    pub fn load_from_message_log(&mut self) -> Result<(), InitError> {
+        while let Some(message) = self.message_log.next_message()? {
+            if let Ok((source, channel)) = self.mqtt_schema.entity_channel_of(&message.topic) {
+                match channel {
+                    Channel::EntityMetadata => {
+                        if let Ok(register_message) = EntityRegistrationMessage::try_from(&message)
+                        {
+                            self.update(register_message)?;
+                        }
+                    }
+                    Channel::EntityTwinData { fragment_key } => {
+                        let fragment_value = if message.payload_bytes().is_empty() {
+                            JsonValue::Null
+                        } else {
+                            serde_json::from_slice::<JsonValue>(message.payload_bytes())?
+                        };
+
+                        self.update_twin_data(EntityTwinMessage::new(
+                            source,
+                            fragment_key,
+                            fragment_value,
+                        ))?;
+                    }
+                    Channel::CommandMetadata { .. } => {
+                        // Do nothing for now as supported operations are not part of the entity store
```

Comment:
Not doing anything here as the supported operations are persisted separately on the file system. A different mechanism already exists to rebuild the entity store from those files. This arm will be enabled once we migrate all that into the entity store.